_posts/2021-02-09-non-key-joins-in-kstreams.adoc

```diff
@@ -0,0 +1,198 @@
+                Materialized.with(Serdes.Long(), addressAndCustomerSerde)
+            )
+            .groupBy(
+                (addressId, addressAndCustomer) -> KeyValue.pair(addressAndCustomer.customer.id, addressAndCustomer),
+                Grouped.with(Serdes.Integer(), addressAndCustomerSerde)
+            )
+            .aggregate(
+                CustomerWithAddresses::new,
+                (customerId, addressAndCustomer, aggregate) -> aggregate.addAddress(addressAndCustomer),
+                (customerId, addressAndCustomer, aggregate) -> aggregate.removeAddress(addressAndCustomer),
+                Materialized.with(Serdes.Integer(), customerWithAddressesSerde)
+            );
+
+            customersWithAddresses.toStream()
+            .to(
+                customersWithAddressesTopic,
+                Produced.with(Serdes.Integer(), customerWithAddressesSerde)
+            );
+----
+
+The Kafka Streams extension is configured via the Quarkus configuration file `application.properties` with the following configuration:
+
+[source,properties]
+----
+customers.topic=dbserver1.inventory.customers
+addresses.topic=dbserver1.inventory.addresses
+customers.with.addresses.topic=customers-with-addresses
+
+quarkus.kafka-streams.bootstrap-servers=localhost:9092
+quarkus.kafka-streams.application-id=kstreams-fkjoin-aggregator
+quarkus.kafka-streams.application-server=${hostname}:8080
+quarkus.kafka-streams.topics=${customers.topic},${addresses.topic}
+
+# streams options
+kafka-streams.cache.max.bytes.buffering=10240
+kafka-streams.commit.interval.ms=1000
+kafka-streams.metadata.max.age.ms=500
+kafka-streams.auto.offset.reset=earliest
+kafka-streams.metrics.recording.level=DEBUG
+kafka-streams.consumer.session.timeout.ms=150
+kafka-streams.consumer.heartbeat.interval.ms=100
+----
+
+== Building and Running the Application
+
+You can now build the aggregator application using:
+----
+mvn clean package -f aggregator/pom.xml
+----
+
+Setup the necessary environment variables:
+----
+export DEBEZIUM_VERSION=1.4
+export QUARKUS_BUILD=jvm
+----
+
+To launch all the containers and building aggregator container images run:
+----
+docker-compose up --build
+----
+
+Configure the Debezium Connector:
+----
+http PUT http://localhost:8083/connectors/inventory-connector/config < register-postgres.json
+----
+
+Now run an instance of the `debezium/tooling` image, attaching to the same network all the other containers run in:
+----
+docker run --tty --rm \
+    --network kstreams-fk-join-network \
+    debezium/tooling:1.1 \
+----
+This image provides several useful tools such as kafkacat. Within the tooling container, run kafkacat to examine the results of the streaming pipeline:
+
+----
+kafkacat -b kafka:9092 -C -o beginning -q \
+    -t customers-with-addresses | jq .
+----
+
+If you want to run this application in `native` mode refer to this https://quarkus.io/guides/kafka-streams#running-natively[detailed guide].
+
+== Summary
```

Comment:
Some more things mentioning to motivate the usage of the extension:

* It exposes health checks
* It awaits all input topics (normally, Kafka Streams will fail if not all topics exist yet), and shows missing ones in the health check
* Metrics, if this is supported out of the box now; needs checking with @ebullient (who works on Metrics integretation into Quarkus via Micrometer) whether that is already present or not
* See whether adding a topology visualization (via https://zz85.github.io/kafka-streams-viz/) is interesting or not